File: common_book/src/txtqbmfxs/java/com/intelligent/reader/adapter/BookShelfItemHolder.java
package com.intelligent.reader.adapter;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.intelligent.reader.R;
import com.intelligent.reader.adapter.holder.AbsRecyclerViewHolder;

import net.lzbook.kit.constants.ReplaceConstants;
import net.lzbook.kit.data.bean.Book;
import net.lzbook.kit.utils.AppUtils;
import net.lzbook.kit.utils.Tools;

import android.content.res.Resources;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

/**
 * Created by Administrator on 2017/4/13 0013.
 */

public class BookShelfItemHolder extends AbsRecyclerViewHolder<Book> {

    private final ImageView mCheck;
    private final ImageView mImage;
    private final TextView mName, mUpdateTime, mUnread, mLastChapter, mUpdate;


    public BookShelfItemHolder(View itemView,
                               BookShelfReAdapter.ShelfItemClickListener shelfItemClickListener,
                               BookShelfReAdapter.ShelfItemLongClickListener shelfItemLongClickListener) {
        super(itemView, shelfItemClickListener, shelfItemLongClickListener);

        mCheck = (ImageView) itemView.findViewById(R.id.check_delete);
        mImage = (ImageView) itemView.findViewById(R.id.book_shelf_image);
        mUpdate = (TextView) itemView.findViewById(R.id.book_shelf_status_update);
        mName = (TextView) itemView.findViewById(R.id.book_shelf_name);
        mUpdateTime = (TextView) itemView.findViewById(R.id.book_shelf_update_time);
        mLastChapter = (TextView) itemView.findViewById(R.id.book_shelf_last_chapter);
        mUnread = (TextView) itemView.findViewById(R.id.book_shelf_unread);

    }

    @Override
    public void onBindData(int position, Book book, boolean update, boolean isRemoveMode, boolean
            removeMark) {
        if (!TextUtils.isEmpty(book.name))
            this.mName.setText(book.name);

        if (book.sequence + 1 > book.chapter_count) {
            book.sequence = book.chapter_count - 1;
        }

        if (book.sequence >= 0) {
            this.mUnread.setText(book.sequence + 1 + "/" + book.chapter_count + "章");
        } else {
            this.mUnread.setText("未读");
        }

//        if (book.sequence == -2) {
//            book.sequence = -1;
//        }
//        int update_count;
//        update_count = book.chapter_count - (book.sequence + 1);
//
//        //快读全本电子书 和 TXT全本小说阅读器 隐藏书架未读 for 差异化上线
//        if ("cn.zsqbydq.reader".equals(AppUtils.getPackageName()) || "cn.kkqbtxtxs.reader".equals(AppUtils.getPackageName())) {
//            this.mUnread.setVisibility(View.GONE);
//        } else {
//            if (update_count == 0) {
//                this.mUnread.setVisibility(View.GONE);
//            } else {
//                this.mUnread.setVisibility(View.VISIBLE);
//                this.mUnread.setText(String.valueOf(update_count) + "章");
//            }
//        }

        // 是否连载
        if (book.status == 2) {
            this.mLastChapter.setText("已完结");
        } else {
            if (!TextUtils.isEmpty(book.last_chapter_name))
                this.mLastChapter.setText(book.last_chapter_name);
        }
        // 是否有更新
        if (!update) {
            this.mUpdate.setVisibility(View.GONE);
        } else {
            this.mUpdate.setVisibility(View.VISIBLE);
        }
        if (this.mUpdateTime != null) {
            this.mUpdateTime.setText(Tools.compareTime(AppUtils.formatter, book
                    .last_updatetime_native));
        }

        if (!TextUtils.isEmpty(book.img_url) && !book.img_url.equals(ReplaceConstants.getReplaceConstants().DEFAULT_IMAGE_URL)) {
            Glide.with(itemView.getContext().getApplicationContext()).load(book.img_url).placeholder(R.drawable.icon_book_cover_default).error((R.drawable.icon_book_cover_default)).diskCacheStrategy(DiskCacheStrategy.ALL).into(this.mImage);
        } else {
            Glide.with(itemView.getContext().getApplicationContext()).load(R.drawable.icon_book_cover_default).into(this.mImage);
        }


        if (isRemoveMode) {
            this.mCheck.setVisibility(View.VISIBLE);
            int typeColor = 0;
            if (removeMark) {
                typeColor = R.mipmap.bookshelf_delete_checked;
            } else {
                typeColor = R.mipmap.bookshelf_delete_unchecked;
            }
            this.mCheck.setBackgroundResource(typeColor);
        } else {
            this.mCheck.setVisibility(View.GONE);
        }
    }
}

File: bookshelf/src/zsmfqbxs/java/com/dingyue/bookshelf/BookShelfItemHolder.kt
package com.dingyue.bookshelf

import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View

import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import android.view.ViewGroup
import com.ding.basic.bean.Book
import net.lzbook.kit.constants.ReplaceConstants
import net.lzbook.kit.utils.AppUtils
import net.lzbook.kit.utils.Tools
import kotlinx.android.synthetic.zsmfqbxs.item_bookshelf_book.view.*
import net.lzbook.kit.app.base.BaseBookApplication
import net.lzbook.kit.constants.Constants
import net.lzbook.kit.utils.book.RepairHelp

/**
 * Desc 书架Item
 * Created by zhenxiang
 * on 2018/5/12 0013.
 */

class BookShelfItemHolder(parent: ViewGroup) : RecyclerView.ViewHolder(
        LayoutInflater.from(parent.context).inflate(R.layout.item_bookshelf_book, parent, false)) {

    fun bind(book: Book, bookshelfItemListener: BookShelfAdapter.BookShelfItemListener,
             contains: Boolean, remove: Boolean) = with(itemView) {
        if (!TextUtils.isEmpty(book.name)) {
            txt_book_name.text = book.name
        }

        if(book.last_chapter != null){
            if (book.sequence + 1 > book.last_chapter!!.serial_number) {
                book.sequence = book.last_chapter!!.serial_number - 1
            }

            if (book.sequence + 1 == book.last_chapter!!.serial_number) {
                txt_book_unread_chapters.visibility = View.GONE
            } else {
                txt_book_unread_chapters.visibility = View.VISIBLE
                if(book.sequence >= 0){
                    txt_book_unread_chapters.text = ((book.sequence + 1).toString() + "/" + book.last_chapter!!.serial_number + "章")
                }else{
                    txt_book_unread_chapters.text = ("0" + "/" + book.last_chapter!!.serial_number + "章")
                }
            }
        }

        /**
         * 书架检测到书籍有修复会在该书籍封面显示更新角标，
         * 并且章节信息变更为：章节已修复至最新（列表书架显示，九宫格书架只显示更新角标）
         * 目录修复：如用户未点击更新弹窗的同步按钮，则书籍封面上的更新角标和更新文案一直存在
         */
        val sp = BaseBookApplication.getGlobalContext().getSharedPreferences(Constants.SHAREDPREFERENCES_KEY, 0)
        if (RepairHelp.isShowFixBtn(context, book.book_id) && sp.getBoolean(book.book_id, true)) {
            img_book_status_update.visibility = View.VISIBLE
        }else{
            when {
                book.update_status == 1 -> { //更新
                    img_book_status_update.visibility = View.VISIBLE
                }
                book.status == "FINISH" -> { //完结
                    img_book_status.visibility = View.VISIBLE
                    img_book_status.setImageResource(R.drawable.bookshelf_item_book_status_finish_icon)
                }
                else -> {
                    img_book_status.visibility = View.GONE
                    img_book_status_update.visibility = View.GONE
                }
            }

        }

        if (txt_book_last_update_time != null && book.last_chapter != null) {
            txt_book_last_update_time.text = (Tools.compareTime(AppUtils.formatter, book
                    .last_chapter!!.update_time) + "更新")
        }

        if ((!TextUtils.isEmpty(book.img_url) && book.img_url != ReplaceConstants.getReplaceConstants().DEFAULT_IMAGE_URL)) {
            Glide.with(itemView.context.applicationContext)
                    .load(book.img_url)
                    .placeholder(R.drawable.book_cover_default)
                    .error(R.drawable.book_cover_default)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(img_book_cover)
        } else {
            Glide.with(itemView.context.applicationContext)
                    .load(R.drawable.book_cover_default)
                    .into(img_book_cover)
        }


        if (remove) {
            img_item_select_state.visibility = View.VISIBLE
            img_book_status.visibility = View.GONE
            img_book_status_update.visibility = View.GONE

            if (contains) {
                img_item_select_state.setImageResource(R.drawable.bookshelf_delete_checked)
            } else {
                img_item_select_state.setImageResource(R.drawable.bookshelf_delete_unchecked)
            }
        } else {
            img_item_select_state.visibility = View.GONE
        }

        rl_main.setOnClickListener {
            bookshelfItemListener.clickedBookShelfItem(book, adapterPosition)
        }

        rl_main.setOnLongClickListener {
            bookshelfItemListener.longClickedBookShelfItem()
        }
    }
}
